### ISSUE TITLE
[torch.compile][fullgraph][dynamic shapes] aten.masked_select.default rejected as “Dynamic shape operator” (eager OK)

### ISSUE THREAD

**Issue by tinywisdom** [pytorch/pytorch#165421]
### 🐛 Describe the bug

A tiny model that builds a boolean mask and calls x.masked_select(mask) runs fine in eager, but fails under torch.compile(fullgraph=True) in Dynamo with:
```
torch._dynamo.exc.Unsupported: Dynamic shape operator
Explanation: Operator `aten.masked_select.default`'s output shape depends on input Tensor data.
Hint: Enable tracing of dynamic shape operators with `torch._dynamo.config.capture_dynamic_output_shape_ops = True`
```

masked_select returns a 1D tensor with length equal to the number of True entries → data-dependent output size. The rejection blocks full-graph compilation of common masking patterns.

### Minimal Repro
```python
import torch, time
from torch import nn

class MyModel(nn.Module):
    def forward(self, x):
        mask = x.ge(0.5)          # boolean mask
        return x.masked_select(mask)

def GetInput():
    return torch.rand(3, 4, dtype=torch.float32).cuda()  # CPU also repros (front-end)

def run_once(model, x, label):
    model.eval()
    t0 = time.perf_counter()
    with torch.no_grad():
        y = model(x)
    print(f"[{label}] output_len={y.numel()}  time={(time.perf_counter()-t0)*1000:.3f}ms")
    return y

def main():
    print("[Env] torch.__version__ =", torch.__version__)
    device = "cuda" if torch.cuda.is_available() else "cpu"
    m = MyModel().to(device)
    x = GetInput().to(device)

    # Eager: OK
    run_once(m, x, "eager")

    # Compile: fails in Dynamo (fullgraph=True)
    cm = torch.compile(m, fullgraph=True)
    run_once(cm, x, "compiled")   # <-- raises Unsupported: Dynamic shape operator

if __name__ == "__main__":
    main()
```

### Error logs

```
[eager] output: [...values...]  # OK
=== torch.compile(fullgraph=True) ===
torch._dynamo.exc.Unsupported: Dynamic shape operator
  Explanation: Operator `aten.masked_select.default`'s output shape depends on input Tensor data.
  Hint: Enable tracing of dynamic shape operators with `torch._dynamo.config.capture_dynamic_output_shape_ops = True`
from user code:
  return x.masked_select(mask)
```

### Versions

```
PyTorch version:  2.8.0a0+gitba56102 
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

cc @chauhang @penguinwu @ezyang @bobrenjc93
